json-rpc/tests/integration_test.rs

```diff
@@ -81,23 +81,40 @@ fn create_test_cases() -> Vec<Test> {
         Test {
             name: "block metadata",
             run: |env: &mut testing::Env| {
-                let resp = env.send("get_metadata", json!([]));
-                let metadata = resp.result.unwrap();
-                assert_eq!(metadata["chain_id"], resp.diem_chain_id);
-                assert_eq!(metadata["timestamp"], resp.diem_ledger_timestampusec);
-                assert_eq!(metadata["version"], resp.diem_ledger_version);
+                // batch request
+                let resp = env.send_request(json!([
+                    {"jsonrpc": "2.0", "method": "get_metadata", "params": [], "id": 1},
+                    {"jsonrpc": "2.0", "method": "get_state_proof", "params": [0], "id": 2}
+                ]));
+
+                // extract both responses
+                let resps: Vec<serde_json::Value> = serde_json::from_value(resp).expect("should be valid serde_json::Value");
+                let metadata = &resps.iter().find(|g| g["id"] == 1).unwrap()["result"];
+                let state_proof = &resps.iter().find(|g| g["id"] == 2).unwrap()["result"];
+
+                // extract header and ensure they match in both responses
+                let diem_chain_id = &resps[0]["diem_chain_id"];
+                let diem_ledger_timestampusec = &resps[0]["diem_ledger_timestampusec"];
+                let diem_ledger_version = &resps[0]["diem_ledger_version"];
+
+                assert_eq!(diem_chain_id, &resps[1]["diem_chain_id"]);
+                assert_eq!(diem_ledger_timestampusec, &resps[1]["diem_ledger_timestampusec"]);
+                assert_eq!(diem_ledger_version, &resps[1]["diem_ledger_version"]);
+
+                // parse metadata
+                assert_eq!(&metadata["chain_id"], diem_chain_id);
+                assert_eq!(&metadata["timestamp"], diem_ledger_timestampusec);
+                assert_eq!(&metadata["version"], diem_ledger_version);
                 assert_eq!(metadata["chain_id"], 4);
                 // for testing chain id, we init genesis with VMPublishingOption#open
                 assert_eq!(metadata["script_hash_allow_list"], json!([]));
                 assert_eq!(metadata["module_publishing_allowed"], true);
                 assert_eq!(metadata["diem_version"], 1);
                 assert_eq!(metadata["dual_attestation_limit"], 1000000000);
-                assert_ne!(resp.diem_ledger_timestampusec, 0);
-                assert_ne!(resp.diem_ledger_version, 0);
+                assert_ne!(diem_ledger_timestampusec, 0);
+                assert_ne!(diem_ledger_version, 0);
 
                 // prove the accumulator_root_hash
-                let sp_resp = env.send("get_state_proof", json!([resp.diem_ledger_version]));
```

Comment:
Actually this line was handling the "state can change inbetween" since it receives the "last known version from client" as being `resp.diem_ledger_version`.
So this is a unnecessary PR.

Reply:
But there's a problem here since we check the provided proof against `metadata["accumulator_root_hash"]` below without batching the queries, agreed.